compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -190,6 +190,330 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
+                "atxmega256c3" => elf::EF_AVR_ARCH_XMEGA6,
+                "atxmega256d3" => elf::EF_AVR_ARCH_XMEGA6,
+                "atxmega384c3" => elf::EF_AVR_ARCH_XMEGA6,
+                "atxmega384d3" => elf::EF_AVR_ARCH_XMEGA6,
+                "atxmega128a1" => elf::EF_AVR_ARCH_XMEGA7,
+                "atxmega128a1u" => elf::EF_AVR_ARCH_XMEGA7,
+                "atxmega128a4u" => elf::EF_AVR_ARCH_XMEGA7,
+                "attiny4" => elf::EF_AVR_ARCH_AVRTINY,
+                "attiny5" => elf::EF_AVR_ARCH_AVRTINY,
+                "attiny9" => elf::EF_AVR_ARCH_AVRTINY,
+                "attiny10" => elf::EF_AVR_ARCH_AVRTINY,
+                "attiny20" => elf::EF_AVR_ARCH_AVRTINY,
+                "attiny40" => elf::EF_AVR_ARCH_AVRTINY,
+                "attiny102" => elf::EF_AVR_ARCH_AVRTINY,
+                "attiny104" => elf::EF_AVR_ARCH_AVRTINY,
+                "attiny202" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny402" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny204" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny404" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny804" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny1604" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny406" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny806" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny1606" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny807" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny1607" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny212" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny412" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny214" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny414" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny814" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny1614" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny416" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny816" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny1616" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny3216" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny417" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny817" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny1617" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny3217" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny1624" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny1626" => elf::EF_AVR_ARCH_XMEGA3,
+                "attiny1627" => elf::EF_AVR_ARCH_XMEGA3,
+                "atmega808" => elf::EF_AVR_ARCH_XMEGA3,
+                "atmega809" => elf::EF_AVR_ARCH_XMEGA3,
+                "atmega1608" => elf::EF_AVR_ARCH_XMEGA3,
+                "atmega1609" => elf::EF_AVR_ARCH_XMEGA3,
+                "atmega3208" => elf::EF_AVR_ARCH_XMEGA3,
+                "atmega3209" => elf::EF_AVR_ARCH_XMEGA3,
+                "atmega4808" => elf::EF_AVR_ARCH_XMEGA3,
+                "atmega4809" => elf::EF_AVR_ARCH_XMEGA3,
+                _ => 0,
```

Comment:
Is this a reasonable fall-back value for an unknown CPU? That is, will the tooling continue working with the objects if this field stays zero?

If not, there are two alternatives to consider here – a compiler error or falling back to a common denominator that is (almost?) guaranteed to work whatever the CPU is specified.

Reply:
> Is this a reasonable fall-back value for an unknown CPU?

I believe zero is the value used for "generic" or "unspecified" in this case. It's also the value that `e_flags` has on AVR without this patch.

For example, in the error message in #106576 (``avr architecture of input file `/tmp/rustcmivxhx/symbols.o' is incompatible with avr:103 output``) , the `103` is the expected `e_flags` value. If the value is zero, then the linker just prints `avr`, like at the beginning of the error message, implying that it's "generic" AVR.

Another example of the same error, but different values: ``avr:51 architecture of input file `main.o' is incompatible with avr output`` ([source](https://stackoverflow.com/questions/51505057/avr-ld-avr51-architecture-of-input-file-main-o-is-incompatible-with-avr-outp))